src/v2/options.ts

```diff
@@ -0,0 +1,270 @@
+  'PRIVATE_RANGES_ONLY',
+  'ALL_TRAFFIC',
+] as const;
+
+/**
+ * A valid VPC Egress setting.
+ */
+export type VpcEgressSetting = typeof SUPPORTED_VPC_EGRESS_SETTINGS[number];
+
+/**
+ * List of available options for IngressSettings.
+ */
+export const SUPPORTED_INGRESS_SETTINGS = [
+  'ALLOW_ALL',
+  'ALLOW_INTERNAL_ONLY',
+  'ALLOW_INTERNAL_AND_GCLB',
+] as const;
+
+export type IngressSetting = typeof SUPPORTED_INGRESS_SETTINGS[number];
+
+/**
+ * GlobalOptions are options that can be set across an entire project.
+ * These options are common to HTTPS and Event handling functions.
+ */
+export interface GlobalOptions {
+  /**
+   * Region where functions should be deployed.
+   * HTTP functions can override and specify more than one region.
+   */
+  region?: SupportedRegion | string;
+
+  /**
+   * Amount of memory to allocate to a function.
+   * A value of null restores the defaults of 256MB.
+   */
+  memory?: MemoryOption | null;
+
+  /**
+   * Timeout for the function in sections, possible values are 0 to 540.
+   * HTTPS functions can specify a higher timeout.
+   * A value of null restores the default of 60s
+   */
+  timeoutSeconds?: number | null;
+
+  /**
+   * Min number of actual instances to be running at a given time.
+   * Instances will be billed for memory allocation and 10% of CPU allocation
+   * while idle.
+   * A value of null restores the default min instances.
+   */
+  minInstances?: number | null;
+
+  /**
+   * Max number of instances to be running in parallel.
+   * A value of null restores the default max instances.
+   */
+  maxInstances?: number | null;
+
+  /**
+   * Number of requests a function can serve at once.
+   * Can only be applied to functions running on Cloud Functions v2.
+   * A value of null restores the default concurrency.
+   */
+  concurrency?: number | null;
+  /**
+   * Connect cloud function to specified VPC connector.
+   * A value of null removes the VPC connector
+   */
+  vpcConnector?: string | null;
+
+  /**
+   * Egress settings for VPC connector.
+   * A value of null turns off VPC connector egress settings
+   */
+  vpcConnectorEgressSettings?: VpcEgressSetting | null;
+
+  /**
+   * Specific service account for the function to run as.
+   * A value of null restores the default service account.
+   */
+  serviceAccount?: string | null;
+
+  /**
+   * Ingress settings which control where this function can be called from.
+   * A value of null turns off ingress settings.
+   */
+  ingressSettings?: IngressSetting | null;
+
+  /**
+   * User labels to set on the function.
+   */
+  labels?: Record<string, string>;
+}
+
+let globalOptions: GlobalOptions | undefined;
+
+/**
+ * Sets default options for all functions written using the v2 SDK.
+ * @param options Options to set as default
+ */
+export function setGlobalOptions(options: GlobalOptions) {
+  if (globalOptions) {
+    logger.warn('Calling setGlobalOptions twice leads to undefined behavior');
```

Comment:
Is this actually true? Looks pretty clear that it just overwrite all previous global options.

Reply:
Undefined doesn't mean unpredictable. I just want to set expectations that this is not an OK thing to do. The warning might also be interesting if someone is surprised that it's being set twice. E.g. If they have multiple files they're importing which each set global options. Or if some malicious codebase decided it would be funny to call `setGlobalOptions({minInstances: 100})`